rfcs/0048-on-conflict-clause-in-create-table-statement.md

```diff
@@ -0,0 +1,121 @@
+  
+### Design
+
+#### Syntax and semantics
+
+```SQL
+  CREATE TABLE [ IF NOT EXISTS ] table_name (
+    col_name data_type [ PRIMARY KEY ],
+    ...
+    [ PRIMARY KEY (col_name, ... ) ]
+  ) ON CONFLICT conflict_action;
+
+  where conflict_action can be one of:
+    DO NOTHING
+    DO OVERWRITE
+    DO UPDATE SET { column_name = expression |
+                    ( column_name [, ...] ) = [ ROW ] ( expression[, ...] )
+                  } [, ...]
+              [ WHERE condition ]
+```
+
+I will compare it with the PG's ON CONFLICT clause(<https://www.postgresql.org/docs/current/sql-insert.html>) in the following text.
+
+```SQL
+INSERT INTO table_name [ AS alias ] [ ( column_name [, ...] ) ]
+    [ ON CONFLICT [ conflict_target ] conflict_action ]
+
+where conflict_target can be one of:
+    ( { index_column_name | ( index_expression ) } [ COLLATE collation ] [ opclass ] [, ...] ) [ WHERE index_predicate ]
+    ON CONSTRAINT constraint_name
+
+and conflict_action is one of:
+    DO NOTHING
+    DO UPDATE SET { column_name = { expression | DEFAULT } |
+                    ( column_name [, ...] ) = [ ROW ] ( { expression | DEFAULT } [, ...] ) |
+                    ( column_name [, ...] ) = ( sub-SELECT )
+                  } [, ...]
+              [ WHERE condition ]
+```
+
+- **Defined when `CREATE TABLE` instead of `INSERT`**
+  Why:
+  1. Different with PG, the conflicts are not only happens in `INSERT` statement but also on the data from external source.
+  2. In PG, users can define different behavior on different `INSERT` statement. We can not implement it easily because RW treat DML statement with just one `Materialize` operator as a streaming job.
+  
+  By the way, for the `UPDATE` statement, we can just override the old value and it can not break the PK constrain after we ban the update on PK columns <https://github.com/risingwavelabs/risingwave/issues/7710>.
+
+- **Not specify `conflict_target`**
+  In RW, we only have PK constrain, so we can just do check on PK constrain. Also if the table does not have pk, user can not specify the ON CONFLICT clause.
+
+- **DO OVERWRITE behavior**
+  just a Syntactic sugar because it is a usual usage. It has the same meaning with `DO UPDATE SET (a, b, c, ...) = (excluded.a, excluded.b, excluded.c, ...)`
+
+- **Sub-query/sub-SELECT in UPDATE**
+  It is too complex and we need implement stream join plan for the `CREATE TABLE`. So I think we need to put it on hold.
+
+### Future possibilities
+
+  We can support the `ON CONFLICT` clause in the Insert statement if we support attach the on conflict description for the chunk in `BatchInsert` executor and `StreamDML` executor to let `Materialize` executor know how to handle the conflict. But the on conflict clause is always needed in create table statement for the table with connector.
```

Comment:
There's still "stream path" between the `DML` and `Materialize`, also `Exchange` if there's user-specified PK, so eventually it seems we need to attach this "description" to each row, which sounds an overkill. 🤔